UI: Source context menu

DDRBoxman opened the pull request:
<img width="1449" alt="screen shot 2018-08-23 at 8 42 32 pm" src="https://user-images.githubusercontent.com/207897/44560426-58c11200-a715-11e8-9da7-099cdc82bb27.png">
<img width="1449" alt="screen shot 2018-08-23 at 8 42 46 pm" src="https://user-images.githubusercontent.com/207897/44560427-59f23f00-a715-11e8-9489-c2189638098f.png">

RytoEX commented on UI/window-basic-main.hpp:
```diff
@@ -675,6 +675,17 @@ private slots:
+	// Source Context Buttons
+	void on_sourcePropertiesButton_clicked();
+	void on_sourceFiltersButton_clicked();
+    void on_sourceRotateRightButton_clicked();
```
Style:  Mixed tabs and spaces here.

DDRBoxman commented:
Note for flipping icon colors for themes
.setCompositionMode(QPainter::RasterOp_NotSource)

DDRBoxman commented:
QString filename = QT_UTF8(":/source/images/sources/") + type + QT_UTF8(".svg");
QSvgRenderer svgRenderer(filename);
QPixmap pix(svgRenderer.defaultSize());
pix.fill(Qt::transparent);
QPainter pixPainter(&pix);
pixPainter.setCompositionMode(QPainter::RasterOp_NotSource); // Flip the colour from black to white
svgRenderer.render(&pixPainter);

popupItem->setIcon(QIcon(pix));

DDRBoxman commented:
<img width="1449" alt="screen shot 2018-08-24 at 9 40 26 pm" src="https://user-images.githubusercontent.com/207897/44614225-5aa6d600-a7e6-11e8-80e7-9d28c7f1e71a.png">

SuslikV commented:
SVG is not the best solution for multiplatform with Qt. There, it is better to use PNG files, I think.

Toolbar should be floating window (or be able to change its position, be hidden). Also my IMHO.

ntoff commented:
Please no. I like to keep my UI as clean and small as possible to fit in 1 corner of the 2nd screen, I remove a few of the docks I already don't use, if this is a thing, please at least make it optional or part of some removable dock. Thanks.

DDRBoxman commented:
<img width="1449" alt="screen shot 2018-08-27 at 12 53 37 am" src="https://user-images.githubusercontent.com/207897/44642981-62a47880-a994-11e8-8f8f-c44c7c497e30.png">

DDRBoxman commented:
<img width="1449" alt="screen shot 2018-08-27 at 1 03 45 am" src="https://user-images.githubusercontent.com/207897/44643186-3ccba380-a995-11e8-9d64-379282cd027b.png">

Heufneutje commented:
I also agree that this should probably be inside of a dock.

dodgepong commented:
Perhaps, but making it a dock will make it take even more vertical room due to the header. If vertical space is such a concern, I'd rather have it with a enable/disable toggle.

It's worth noting that one of the biggest motivating forces for this feature is to have a place to put media controls without having to open a properties window.

Heufneutje commented:
Well, for me it's not really the vertical space that's an issue. It's more wanting to position it where I like it, the same way the other docks work.

ntoff commented:
Even with media controls there, that's just more wasted space in my eyes, will those also be optionally shown? I like to make the obs window as small and compact as possible to fit a few things on my 2nd screen, I don't need media controls or some stupid buttons taking up precious space, if I could disable the video preview window I would because I don't even need that (all my scenes and overlays are handled on another PC, OBS is nothing more than a glorified mixer / video sender to me but I still need to see that mixer and the volume levels).

dodgepong commented:
> I still need to see that mixer and the volume levels

So, I assume that means you pop-out the mixer and then just hide the rest of the window underneath other things?

ntoff commented:
No, I have it all sort of as small as possible so I can see the mixer, and whether or not I'm on the "games" scene of "system" scene.

![image](https://user-images.githubusercontent.com/9030085/44702648-bc618b80-aad7-11e8-8928-6330abd022af.png)

I have it sitting in 1 corner of the screen, making sure my mic & other things are bouncing the little meters around, and browser window(s) in another corner(s) and perhaps voicemeeter in another corner.

I wasn't even aware you *could* pop things out (I only knew about dragging them around), this pleases us.

This is probably the most common way I have things organised 

![image](https://user-images.githubusercontent.com/9030085/44702776-64775480-aad8-11e8-8a52-3451bb899182.png)

So having less stuff cluttering up the UI the more room there is for the mixer, or the buttons to see whether the replay buffer is going, whether I'm on the games or system scene (ignore all the other scenes, sometimes I add stuff just to test, the only 2 that get used are games & system).

I guess if worst comes to worst I'll just pop everything out and have a bunch of random floating windows around like the GiMP had.

RytoEX commented:
@Heufneutje 
> Edit: I would like to mention that I definitely like the idea behind this and I understand why it's a good idea to have it. I'm just not sure I quite like it's fixed location. That's all.

As far as I'm aware, making it dockable or making it able to be hidden/shown are things that are being considered or looked into.  No promises on the final implementation, as this is a work-in-progress.

@ntoff 
> Even with media controls there, that's just more wasted space in my eyes, will those also be optionally shown?

Media Controls would likely be optional, if implemented this way.  This is very much still a work-in-progress towards broader goals of UI improvements to better enable feature discovery and spending less time in right-click menus and property windows.

> if I could disable the video preview window I would

You can.  Right-click the preview and uncheck "Enable Preview".  It will still take up some space, but not quite as much as with the preview enabled.
![image](https://user-images.githubusercontent.com/624931/44702748-ed53a900-aa62-11e8-8154-aa00d2856c15.png)


You also detach the Mixer dock instead, as @dodgepong said.  You can also close or detach any of the other docks (Scenes, Sources, Scene Transitions, Controls).  While minimizing OBS will also minimize detached docks, you _can_ hide OBS to the system tray instead and detached docks will remain visible (though such docks will become unable to be targeted by alt-tab on Windows).
![image](https://user-images.githubusercontent.com/624931/44702752-f3498a00-aa62-11e8-831c-05b3289a013b.png)

Or make the mixer smaller:
![image](https://user-images.githubusercontent.com/624931/44702963-c9dd2e00-aa63-11e8-966d-410cc964d313.png)

Or with mixer's new Vertical Layout:
![image](https://user-images.githubusercontent.com/624931/44702984-dceffe00-aa63-11e8-9b1d-4caf42e698f6.png)


@DDRBoxman 
Since there's clearly a lot to talk about and consider on this, and things are still in motion, perhaps retitle the PR with a "WIP" prefix?

RytoEX commented:
@ntoff 
You and I cross-posted.  I was still typing my reply when yours came in, and I hadn't checked for new comments before submitting mine, so my comment above was made without the context of [your latest comment](https://github.com/obsproject/obs-studio/pull/1447#issuecomment-416458017).

ntoff commented:
>You can. Right-click the preview and uncheck "Enable Preview". It will still take up some space, but not quite as much as with the preview enabled.

Oops I meant disable as in completely remove (a question I've asked before and know to be impossible), I know you can "turn it off" which I already do (I enable it, make sure the game is being picked up, then disable it usually).

Warchamp7 commented:
For the sake of discussion, these are the mockups I created when proposing this idea. The big thing as mentioned was having a way to expose media source controls in a way that would also be useful for other sources.

![item controls](https://user-images.githubusercontent.com/1554753/44749395-5cbbae00-aae0-11e8-90e5-251e074060a5.jpg)

![v23 ui bottom](https://user-images.githubusercontent.com/1554753/44749537-c20f9f00-aae0-11e8-92fd-c125135a3af8.jpg)

This is the space I would expect the mac screenshot to take up
![mac prototype mockup](https://user-images.githubusercontent.com/1554753/44749547-ce93f780-aae0-11e8-8926-67be87a3eaa6.png)

DDRBoxman commented:
<img width="610" alt="screen shot 2018-09-07 at 10 25 33 pm" src="https://user-images.githubusercontent.com/207897/45249839-096e0a80-b2ed-11e8-9626-927d6a7027c1.png">

Needs signals hooked up for the new checkboxes

DDRBoxman commented:
Okay, I went ahead and rebased this and removed the hide / lock / name parts. I was having a hard time figuring out a good way to deal with the signal binding / unbinding to make those work.

WizardCM commented:
* I'd vote for less spacing above/below.
* It's not immediately obvious why the menu is missing when a locked source is selected
* It seems to work backwards when using the preview to select sources? I seem to be getting a context menu when nothing is selected, then losing it when my 1 source is selected. Using the Sources list works correctly.

DDRBoxman commented:
![contextsmaller](https://user-images.githubusercontent.com/207897/54004830-bd294900-411c-11e9-9615-cfefdffa1dbe.PNG)

DDRBoxman commented:
<img width="1191" alt="Screen Shot 2019-03-12 at 6 41 57 PM" src="https://user-images.githubusercontent.com/207897/54243466-8e351d80-44f6-11e9-83d4-f22c43ddc0a1.png">
<img width="1191" alt="Screen Shot 2019-03-12 at 6 42 00 PM" src="https://user-images.githubusercontent.com/207897/54243467-8e351d80-44f6-11e9-8b21-ff768f4c2973.png">
